specs/waku/v2/waku-store.md

```diff
@@ -2,36 +2,88 @@
 
+## Waku Store Protocol 
 `WakuStore` is a protocol to enable querying of messages received through relay protocol and stored by other nodes.
 
+## Pagination
+Nodes supporting store protocol MAY support Pagination. When a node supports pagination, a requester node can set the number of waku messages to return and the offset to start from. The queried node MUST then use this information to return messages in the queried range. The details of the pagination are explained next.
+
+To perform pagination, each `WakuMessage` stored at a waku store node is associated with a unique `Index`.  Once a `WakuMessage` is received by a node, it will generate the `Index` of the  `WakuMessage` and then saves the (`Index`, `WakuMessage`) together in its local storage. An `Index` encapsulates fields useful for paging, these fields are a `digest` and a `receivedTime`. The `receivedTime` indicates the time at which the `WakuMessage` is received by the store node and the `digest` is the hash of the received `WakuMessage`.  
+
+
+
+When a `HistoryQuery` arrives, it may contain an extra field `pagingInfo` that carries the information required for paging. The `pagingInfo` contains the `pageSize`, `cursor` and the `direction` as defined below:
+
+**pageSize**
+
+- A positive integer indicating the number of `WakuMessage`s to be included in the `HistoryResponse`
+- If the `pageSize` is omitted then a “default page size” shall be used
+- There is an upper limit for the page size
+- If the number of remaining `WakuMessage`s is less than the `pageSize`, then the `pageSize` is adjusted based on the remaining messages
+
+**Cursor**
+
+-  Holds the `Index` of a waku message
+
+**Direction**
+
+- Is a boolean value
+     - True means forward pagination that is the `messages` field of the `HistoryResponse` contains `pageSize` many (or less) waku messages whose `Index` are larger than the given `cursor`
+     - False means backward pagination, that is the `messages` field of the `HistoryResponse` contains  `pageSize` many (or less) waku messages whose `Index` are smaller than the given cursor
+
+Given a valid  `pagingInfo`, the waku store node performs as follows:
+- sorts the list of `WakuMessage`s based on their `Index`
+- finds the `WakuMessage` whose `Index` is the same as `cursor` 
+- Extracts `pageSize` many waku messages whose `Index` exceed/precede the `cursor` (depending on the `direction` being true/false)
+- the resultant list  is sent in the  `messages` field of the `HistoryResponse`
+
+The `HistoryResponse` will also contain an updated `pagingInfo` which helps the requester to resume its further requests; the `pageSize` indicates the number of returned waku messages (i.e., the number of messages included in `messages` field of `HistoryResponse`), the `direction` is the same direction as in the corresponding `HistoryQuery`. The `cursor` holds the `Index` of the last/first message in the `HistoryResponse` `messages` (depending on the `direction` being true/false).
```

Comment:
For true/false statements, I'd probably write it explicitly with one case, and then write something like "...and vice versa for direction being false" or "...and true and false respectively for foo".

Reply:
Might want to add a sentence or paragraph on how this allows a querying node to receive all messages in an iterated manner, either starting from the beginning or from the end.

Also, probably belongs further up as a general comment, all of this assumes the node was online to receive messages from relay protocol. There could still be gaps :<

Reply:
By an online node you meant the responder node is supposed to be online and receive all the messages, right?Then sure, gaps are inevitable, indeed we never attempted this problem (gaps) in the current pagination method, the current design only works for one requester and responder. Please let me know if a separate issue is needed to address the offline (store) nodes?

Reply:
Yeah, I think just a sentence or so to make it clear to the reader what the assumption is, since it might not be obvious to everyone.

I think there might already be an issue for the general store high availability requirement problem, (@decanus might know), but feel free to create one if you want! Definitely not a part of the scope of the current work though, as you say.